src/main/drivers/dshot_bb.c

```diff
@@ -0,0 +1,697 @@
+        return NULL;
+    }
+
+    gpioToMotorPort[portIndex] = bbPort;
+    ++usedMotorPorts;
+
+    return bbPort;
+}
+
+// Return frequency of smallest change [state/sec]
+
+uint32_t getDshotBaseFrequency(motorPwmProtocolTypes_e pwmProtocolType)
+{
+    switch (pwmProtocolType) {
+    case(PWM_TYPE_PROSHOT1000):
+        return MOTOR_PROSHOT1000_SYMBOL_RATE * MOTOR_PROSHOT1000_STATE_PER_SYMBOL;
+    case(PWM_TYPE_DSHOT1200):
+        return MOTOR_DSHOT1200_SYMBOL_RATE * MOTOR_DSHOT_STATE_PER_SYMBOL;
+    case(PWM_TYPE_DSHOT600):
+        return MOTOR_DSHOT600_SYMBOL_RATE * MOTOR_DSHOT_STATE_PER_SYMBOL;
+    case(PWM_TYPE_DSHOT300):
+        return MOTOR_DSHOT300_SYMBOL_RATE * MOTOR_DSHOT_STATE_PER_SYMBOL;
+    default:
+    case(PWM_TYPE_DSHOT150):
+        return MOTOR_DSHOT150_SYMBOL_RATE * MOTOR_DSHOT_STATE_PER_SYMBOL;
+    }
+}
+
+static void bbAllocDMA(bbPort_t *bbPort)
+{
+    const timerHardware_t *timhw = bbPort->timhw;
+
+#ifdef USE_DMA_SPEC
+    dmaoptValue_t dmaopt = dmaGetOptionByTimer(timhw);
+    const dmaChannelSpec_t *dmaChannelSpec = dmaGetChannelSpecByTimerValue(timhw->tim, timhw->channel, dmaopt);
+    bbPort->dmaResource = dmaChannelSpec->ref;
+    bbPort->dmaChannel = dmaChannelSpec->channel;
+#else
+    bbPort->dmaResource = timhw->dmaRef;
+    bbPort->dmaChannel = timhw->dmaChannel;
+#endif
+
+    dmaIdentifier_e dmaIdentifier = dmaGetIdentifier(bbPort->dmaResource);
+    dmaInit(dmaIdentifier, OWNER_BBSHOT, RESOURCE_INDEX(bbPort - bbPorts));
+    bbPort->dmaSource = timerDmaSource(timhw->channel);
+
+    bbPacer_t *bbPacer = bbFindMotorPacer(timhw->tim);
+    bbPacer->dmaSources |= bbPort->dmaSource;
+
+    dmaSetHandler(dmaIdentifier, bbDMAIrqHandler, NVIC_BUILD_PRIORITY(2, 1), (uint32_t)bbPort);
+
+    bbDMA_ITConfig(bbPort);
+}
+
+void bbDMAIrqHandler(dmaChannelDescriptor_t *descriptor)
+{
+    dbgPinHi(0);
+
+    bbPort_t *bbPort = (bbPort_t *)descriptor->userParam;
+
+    bbDMA_Cmd(bbPort, DISABLE);
+
+    bbTIM_DMACmd(bbPort->timhw->tim, bbPort->dmaSource, DISABLE);
+
+    if (DMA_GET_FLAG_STATUS(descriptor, DMA_IT_TEIF)) {
+        while (1) {};
+    }
+
+    DMA_CLEAR_FLAG(descriptor, DMA_IT_TCIF);
+
+#ifdef USE_DSHOT_TELEMETRY
+    if (useDshotTelemetry) {
+        if (bbPort->direction == BBSHOT_DIRECTION_INPUT) {
+#ifdef DEBUG_COUNT_INTERRUPT
+            bbPort->inputIrq++;
+#endif
+        } else {
+#ifdef DEBUG_COUNT_INTERRUPT
+            bbPort->outputIrq++;
+#endif
+
+            // Switch to input
+
+            bbSwitchToInput(bbPort);
+
+            bbTIM_DMACmd(bbPort->timhw->tim, bbPort->dmaSource, ENABLE);
+        }
+    }
+#endif
+    dbgPinLo(0);
+}
+
+// Setup bbPorts array elements so that they each have a TIM1 or TIM8 channel
+// in timerHardware array for BB-DShot.
+
+static void bbFindPacerTimer(void)
+{
+    const timerHardware_t *timer;
+    bool usedTimerChannels[32][CC_CHANNELS_PER_TIMER] = {0};
+
+    for (int bbPortIndex = 0; bbPortIndex < MAX_SUPPORTED_MOTOR_PORTS; bbPortIndex++) {
```

Comment:
I did that for targets like this:
```
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM8, CH3, PC8, TIM_USE_CAMERA_CONTROL, 0, 1), // USED FOR CAMERA CONTROL
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM1, CH3, PE13, TIM_USE_NONE,          0, 1 ), // RC1 / PPM, unusable
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM1, CH1, PE9,  TIM_USE_MOTOR,         0, 2 ), // M3
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM1, CH2, PE11, TIM_USE_MOTOR,         0, 1 ), // M4
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM8, CH1, PC6,  TIM_USE_NONE,          0, 0 ), // UART6_TX
src/main/target//OMNIBUSF7/target.c:    DEF_TIM(TIM8, CH2, PC7,  TIM_USE_NONE,          0, 1 ), // UART6_RX
```

As you see both TIM1 and TIM8 are made available in this target, but for uses that are likely fairly rare (at least both together). Picking dynamically what's still unused allows these targets to work in the vast majority of all cases.

Reply:
Ah, so this only applies to targets that use TIM1 and TIM8 as pin based timers? In this case the potential for conflict is much smaller, and doing this dynamically should be fine.

What happens if there aren't enough timers available, how is this indicated to the user?

Reply:
<img width="755" alt="image" src="https://user-images.githubusercontent.com/18134492/64071559-0518cb00-cc7d-11e9-89d7-06f008b01d5d.png">

Yeah - we can only use timer 1 and timer 8 as gpio bitbang pacers since they are the only  timers that trigger dma streams on dma controller 2 which is the only dma controller that can access APB2 where the gpio IO registers reside (see picture, it's the same on stm32f7).

If no pacing timer can be found the quad will not arm since there's no telemetry available.

Reply:
> If no pacing timer can be found the quad will not arm since there's no telemetry available.

What happens if Dshot telemetry is set to off (while `dshot_bitbang` is `on`)?

Probably better to fail obvious in this case by directly setting arming disabled. For every other function we have, it is simple and obvious that if a resource that is configured as required but not claimed it won't work. With this, since we are automatically allocating resources, it will be hard for the user to determine if this should be working or not.

Reply:
Yes true and good point.

Reply:
done